### ISSUE TITLE
Question about anchor designing

### ISSUE THREAD

**Issue by lidehuihxjz** [ultralytics/yolov5#1344]
## ❔Question
Is the anchor related to the model input size? 
For exsample, my training data is 640 X 320, and there is an anchor of 40 X 30. If model input size is set to 640 X 640, should this anchor be changed to 40 X 60?  
Thanks!

**Comment by glenn-jocher** [comment]
@lidehuihxjz YOLOv5 autoanchor automatically recomputes your anchors as necessary before training starts. No action is required on your part.

**Comment by planemanner** [comment]
@glenn-jocher As you mentioned, if auto-anchor can provide best-fit anchors on custom dataset, how can get the anchors externally?

**Comment by KangHoyong** [comment]
It is said to automatically convert anchors, so does that count in the CPU area or in the CUDA area?

**Comment by glenn-jocher** [comment]
@planemanner anchors are automatically integrated with your model, no action is needed on your part.

**Comment by lidehuihxjz** [comment]
@glenn-jocher Thanks your reply! 
I have one more question here. Since the anchors are automatically computed, why  have the parameter of _anchor_ in yaml file? This parameter seems to be set manually.

**Comment by glenn-jocher** [comment]
@lidehuihxjz autoanchor handles everything. It compares your manual anchors (if supplied) to your data and training arguments, and will evolve new anchors if the supplied ones are a poor fit.

**Comment by lidehuihxjz** [comment]
@glenn-jocher 
I see. You gave a clear explanation! Thanks again!

**Comment by lidehuihxjz** [comment]
@glenn-jocher 
If I used trained model to make the prediction, the evolved anchors will be automatically used. Right?
If so, there may be somewhere to store the evolved anchors. How can I find them?

**Comment by glenn-jocher** [comment]
@lidehuihxjz the anchors are parameters of the Detect() layer. Search the issues, the answer's been posted before.

**Comment by planemanner** [comment]
> @planemanner anchors are automatically integrated with your model, no action is needed on your part.

@glenn-jocher actually, I want to use the automatic algorithm to another detection network. So I questioned about a method which can get anchors externally.

**Comment by lidehuihxjz** [comment]
> @lidehuihxjz the anchors are parameters of the Detect() layer. Search the issues, the answer's been posted before.

I still don't know how to acquire the auto-calculated anchors.  Could you please gave more guidance?

**Comment by lidehuihxjz** [comment]
> @lidehuihxjz the anchors are parameters of the Detect() layer. Search the issues, the answer's been posted before.

I still don't know how to acquire the auto-calculated anchors. Could you please give more guidance? @glenn-jocher

**Comment by lidehuihxjz** [comment]
> @lidehuihxjz the anchors are parameters of the Detect() layer. Search the issues, the answer's been posted before.

I found something else strange. I trianed with size fo 160x256, and used detect.py to do test with size of 256x480. I got the correct predictions without changing the anchors!!!
So, the only possible reason is that anchors are stored with relative value, but the anchor parameter in Detect() layer is absolute value. Why? @glenn-jocher

**Comment by zhiqwang** [comment]
The `Anchor` is an absolute value in most object detection model, so @glenn-jocher set `self.register_buffer(...)` as belows (I guess):

https://github.com/ultralytics/yolov5/blob/201bafc7cf9545362552ad5b0fc5033d64d7ae77/models/yolo.py#L39-L40

BTW, In other model like torchvision's `retinanet` or `faster-rcnn`, their anchor are changed with the shape of input images. After slightly refactoring, we can visulize the model graph as following, it will be easy to explore the model structure.

<a href="https://user-images.githubusercontent.com/4210061/99934857-30b03580-2d9a-11eb-83bc-f02acd12d293.png"><img src="https://user-images.githubusercontent.com/4210061/99934857-30b03580-2d9a-11eb-83bc-f02acd12d293.png" alt="YOLO model visualize" width="500"/></a>

vs

<a href="https://user-images.githubusercontent.com/4210061/99934943-7a008500-2d9a-11eb-83c0-47fffea3fbba.png"><img src="https://user-images.githubusercontent.com/4210061/99934943-7a008500-2d9a-11eb-83c0-47fffea3fbba.png" alt="YOLO model visualize" width="500"/></a>

Check for more details in [my notebooks](https://github.com/zhiqwang/yolov5-rt-stack/blob/master/notebooks/visualize-jit-models.ipynb)

**Comment by glenn-jocher** [comment]
@lidehuihxjz anchor sizes in absolute units of pixels. They are not a function of the input image size.

@zhiqwang oh that's interesting. Yes the two anchor parameters you see there are the same values saved with different shapes for faster speed (one fixed shape for inference, one fixed shape for loss function). They are buffers so they get moved between devices like parameters with .to(device), but are not parameters and thus do not have gradients.

**Comment by glenn-jocher** [comment]
@lidehuihxjz also you will get best results training at the same size you will be running inference at in the future.
